crates/bifrost/src/background_appender.rs

```diff
@@ -0,0 +1,403 @@
+            tokio::select! {
+                _ = &mut drain_fut => {
+                    trace!("Draining the background appender");
+                    // stop accepting messages and drain the queue
+                    rx.close();
+                }
+                received = rx.recv_many(&mut recv_buffer, max_batch_size) => {
+                    if received == 0 {
+                        // channel is closed, appender is drained
+                        break;
+                    }
+
+                    // The background appender stops if a batch write failed. All enqueued messages
+                    // will be dropped and senders will receive [`EnqueueError::Closed`]
+                    //
+                    // All buffers get reset within `process_appends()`
+                    Self::process_appends(
+                        &mut appender,
+                        &mut recv_buffer,
+                        &mut notif_buffer,
+                        &mut batch_buffer,
+                    ).await?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn process_appends(
+        appender: &mut Appender,
+        buffered_records: &mut Vec<AppendOperation<T>>,
+        notif_buffer: &mut Vec<oneshot::Sender<()>>,
+        batch_buffer: &mut Vec<(Bytes, Keys)>,
+    ) -> Result<()> {
+        batch_buffer.reserve(buffered_records.len());
+
+        // Attempt to reserve enough bytes for the payloads
+        appender
+            .serde_buffer
+            .reserve(buffered_records.len() * RECORD_SIZE_HINT);
+        for record in buffered_records.drain(..) {
+            match record {
+                AppendOperation::Enqueue(record) => {
+                    let keys = record.record_keys();
+                    StorageCodec::encode(&record, &mut appender.serde_buffer)
+                        .expect("record serde is infallible");
+                    // todo, optimize to avoid copying the payload bytes again
+                    let payload = Payload::new(appender.serde_buffer.split().freeze());
+                    StorageCodec::encode(payload, &mut appender.serde_buffer)
+                        .expect("record serde is infallible");
+                    info!("payload size ={}", appender.serde_buffer.len());
+                    batch_buffer.push((appender.serde_buffer.split().freeze(), keys));
+                }
+                AppendOperation::EnqueueWithNotification(record, tx) => {
+                    let keys = record.record_keys();
+                    StorageCodec::encode(&record, &mut appender.serde_buffer)
+                        .expect("record serde is infallible");
+                    // todo, optimize to avoid copying the payload bytes again
+                    let payload = Payload::new(appender.serde_buffer.split().freeze());
+                    StorageCodec::encode(payload, &mut appender.serde_buffer)
+                        .expect("record serde is infallible");
```

Comment:
Maybe factor out into a function `encode_as_payload<T: StorageEncode + HasRecordKeys>(value: &T, ...)`? I think I have seen this snippet at a few places.

Reply:
Good idea but this code gets removed in a follow-up PR anyway so I wouldn't worry so much about it.